### PR TITLE
fix: interpolate gen fix (#12716)

### LINKED ISSUE #12716: Interpolated Generator can perform redundant parameter generation
Checklist:

* [x] I've searched in the docs and FAQ for my answer: https://bit.ly/argocd-faq.
* [x] I've included steps to reproduce the bug.
* [x] I've pasted the output of `argocd version`.

**Describe the bug**

We noticed slowdown with our Appsets when upgrading from v2.4.14 -> v2.5.10. We weren't entirely sure what the issue was, but we pinned it down to a feature introduced in [PR-9080](https://github.com/argoproj/argo-cd/pull/9080) (actually contributed by myself, smh).

*Context*
We use many appsets that have a matrix generator composed of a cluster generator + git generator. They would look something like this:

```yaml
apiVersion: argoproj.io/v1alpha1
kind: ApplicationSet
metadata:
  name: cluster-git
spec:
  generators:
    - matrix:
        generators:
          - clusters:
              selector:
                matchLabels:
                  argocd.argoproj.io/secret-type: cluster
                  kubernetes.io/environment: 'dev'
          - git:
              repoURL: https://github.com/argoproj/applicationset.git
              revision: HEAD
              files:
                - path: "examples/git-generator-files-discovery/cluster-config/config.json"
```

What we noticed looking through the appset code was that the time it took the Matrix generator to generate its params was far greater in v2.5.10 versus v2.4.14. For example, here are the top 10 longest times for both versions:

*Matrix-Generator - GenerateParam Times - v2.5.10*
```yaml
 8.649865
 10.368155
 10.881855
 11.163029
 25.036734
 26.115226
 26.382355
 26.935813
 27.704323
 28.153638
 28.374578
```

*Matrix-Generator - GenerateParam Times - v2.4.14*
```yaml
 2.936658   
 2.939816   
 2.939872   
 2.944849   
 2.944878   
 2.948442   
 2.991007   
 2.991060   
 2.991501   
 3.036504   
 3.118068 
```

*TLDR/The Real Issue*
We dug further into what was causing the long reconcile times for the Matrix Generator in v2.5.10. Despite the fact we do not use any interpolated parameters at the moment, we're still noticing multiple `GenerateParam` calls being run in a single Matrix `GenerateParam` call. This is due to the 2nd param being run [within a for loop](https://github.com/argoproj/argo-cd/blob/352e51fdbd1bc729178ca5c16880dc9fcb5a5504/applicationset/generators/matrix.go#L54). 

**To Reproduce**
I created a test-case for each branch to compare the difference in calls. Please see here:

- [v2.5.10](https://github.com/KojoRising/argo-cd/blob/feat/appset_bug_v2.5.10/applicationset/generators/matrix_test.go#L839)

- [v2.4.14](https://github.com/KojoRising/argo-cd/blob/feat/appset_bug_v2.4.14/applicationset/generators/matrix_test.go#L271)

**Expected behavior**

v2.4.14 - Git's GenerateParam should only be called once
image

v2.5.10 - Git's GenerateParam should also only be called once
image

**Screenshots**

**Version**

v2.5.10

**Logs**

Please see above.

### PR DESCRIPTION
**Pull request by KojoRising**
Closes [#12716](https://github.com/argoproj/argo-cd/issues/12716). 

This is a simple fix that does the following: 
- Checks to see if parameter generation for the 2nd Nested Appset Generator can be successfully executed; if so, it skips the for-loop generation that can cause many redundant "GenerateParam" calls
- If no params are successfully generated from above, it proceeds to the for-loop (which performs "GenerateParam" calls for each param from the first Nested Appset Generator). 

Note on DCO:

If the DCO action in the integration test fails, one or more of your commits are not signed off. Please click on the *Details* link next to the DCO action for instructions on how to resolve this.

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [ ] Optional. My organization is added to USERS.md.
* [ ] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)). 
* [x] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [x] I have added a brief description of why this PR is necessary and/or what this PR solves.

Please see [Contribution FAQs](https://argo-cd.readthedocs.io/en/latest/developer-guide/faq/) if you have questions about your pull-request.

### CODE REVIEW

File: applicationset/generators/matrix.go [hunk 1]
```diff
@@ -79,6 +100,63 @@ func (m *MatrixGenerator) GenerateParams(appSetGenerator *argoprojiov1alpha1.App
 	return res, nil
 }
 
+func (m *MatrixGenerator) checkSecondGeneratorForParams(g1 []map[string]interface{}) (bool, error) {
+	for _, a := range g1 {
+		aStr := utils.ConvertToMapStringString(a)
+		for _, v := range aStr {
+			willComp := regexp.MustCompile("\\{\\{.*\\}\\}")
```
Comment (crenshaw-dev) [inline applicationset/generators/matrix.go:107]: Maybe compile at global scope for speeds?

File: applicationset/generators/matrix.go [hunk 2]
```diff
@@ -50,10 +50,17 @@ func (m *MatrixGenerator) GenerateParams(appSetGenerator *argoprojiov1alpha1.App
 	if err != nil {
 		return nil, err
 	}
+	requiresInterpolation := false // Try to generate 2nd generator's params without interpolation
+	g1, err := m.getParams(appSetGenerator.Matrix.Generators[1], appSet, nil)
+	if err != nil || g1 == nil {
```
Comment (crenshaw-dev) [inline applicationset/generators/matrix.go:55]: A little unusual, but I like the simplicity. :-) Wanna add tests to show that the generators aren't run extra times?

**Review by crenshaw-dev — CHANGES_REQUESTED** [review]

File: applicationset/generators/matrix.go (hunk 2 above)
Comment (KojoRising) [reply applicationset/generators/matrix.go:55]: Hey @crenshaw-dev thanks for the review!! Yes for sure, I'm currently in the middle of writing some tests that represent just that.

Comment (KojoRising) [reply applicationset/generators/matrix.go:55]: @crenshaw-dev - The test should be good to go. The test was initially causing many issues for some reason, but came to the realisation that for the "Interpolated Generator" (ie. like below), the 1st `GenerateParams` would end up returning nil/no-op (which makes sense according to the code). So managed to get that factored into the test-case.

```yaml
	interpolatedGitGenerator := &argoprojiov1alpha1.GitGenerator{
		RepoURL:  "RepoURL",
		Revision: "Revision",
		Files: []argoprojiov1alpha1.GitFileGeneratorItem{
			{Path: "examples/git-generator-files-discovery/cluster-config/{{name}}.json"},
		},
	}
```

Comment (KojoRising) [reply applicationset/generators/matrix.go:55]: ## Test Summary

### Test-Setup
Matrix generator w/:
1) Cluster-Generator - dev + prod Clusters
2) Git Generator - With single git-files path entry 

#### Test 1 - Interpolated Generator
- 3 "GenerateParams" calls total - 1 No-Op, 1-Dev, 1-Prod

#### Test 2 - NonInterpolated Generator
- 1 "GenerateParams" call - 1 regular call


Anyways, let me know if you have any feedback/recommendations. Thanks!

Comment (KojoRising) [reply applicationset/generators/matrix.go:55]: @crenshaw-dev - Let me know if there's anything else you need on the PR. Thanks!

**Review by ishitasequeira — APPROVED** [review]
Thanks @KojoRising for the PR. I tested the PR and it does work as described and expected.

**Review by crenshaw-dev — APPROVED** [review]
Thanks,  @KojoRising!

**Comment by KojoRising** [comment]
Thanks again for the review & merge @ishitasequeira and @crenshaw-dev!! Appreciate it!!

**Comment by blakepettersson** [comment]
@KojoRising `TestMatrixGenerate/happy_flow_-_generate_params` is failing on CI on a semi-regular basis, would you happen to have insights as to why that is?

File: applicationset/generators/matrix_test.go [hunk 3]
```diff
@@ -1005,6 +1006,273 @@ func TestMatrixGenerateListElementsYaml(t *testing.T) {
...
+					"name":                        "prod-01",
+					"nameNormalized":              "prod-01",
+					"server":                      "https://prod-01.example.com",
+					"metadata.labels.environment": "prod",
+					"metadata.labels.argocd.argoproj.io/secret-type": "cluster",
+				},
+			},
+			clientError: false,
+			expectedMock: [][]map[string]interface{}{
+				{}, // Needed, because the 1st call will fail without interpolation
```
Comment (crenshaw-dev) [inline applicationset/generators/matrix_test.go:1142]: I think this is true if and only if the generator contains a go template. The non-existence of a parameter does not cause `GenerateParams` to return nil or an error for fasttemplate fields. 

This problem is causing a test failure after another PR was merged:

```
--- FAIL: TestMatrixGenerate (0.00s)
    --- FAIL: TestMatrixGenerate/happy_flow_-_generate_params (0.00s)
        matrix_test.go:190: 
            	Error Trace:	/home/runner/work/argo-cd/argo-cd/applicationset/generators/matrix_test.go:190
            	Error:      	Not equal: 
            	            	expected: []map[string]interface {}{map[string]interface {}{"cluster":"Cluster", "path":"app1", "path.basename":"app1", "path.basenameNormalized":"app1", "templated":"test-app1", "url":"Url"}, map[string]interface {}{"cluster":"Cluster", "path":"app2", "path.basename":"app2", "path.basenameNormalized":"app2", "templated":"test-app2", "url":"Url"}}
            	            	actual  : []map[string]interface {}{map[string]interface {}{"cluster":"Cluster", "path":"app1", "path.basename":"app1", "path.basenameNormalized":"app1", "templated":"test-{{path.basenameNormalized}}", "url":"Url"}, map[string]interface {}{"cluster":"Cluster", "path":"app2", "path.basename":"app2", "path.basenameNormalized":"app2", "templated":"test-{{path.basenameNormalized}}", "url":"Url"}}
            	            	
            	            	Diff:
            	            	--- Expected
            	            	+++ Actual
            	            	@@ -6,3 +6,3 @@
            	            	   (string) (len=23) "path.basenameNormalized": (string) (len=4) "app1",
            	            	-  (string) (len=9) "templated": (string) (len=9) "test-app1",
            	            	+  (string) (len=9) "templated": (string) (len=32) "test-{{path.basenameNormalized}}",
            	            	   (string) (len=3) "url": (string) (len=3) "Url"
            	            	@@ -14,3 +14,3 @@
            	            	   (string) (len=23) "path.basenameNormalized": (string) (len=4) "app2",
            	            	-  (string) (len=9) "templated": (string) (len=9) "test-app2",
            	            	+  (string) (len=9) "templated": (string) (len=32) "test-{{path.basenameNormalized}}",
            	            	   (string) (len=3) "url": (string) (len=3) "Url"
            	Test:       	TestMatrixGenerate/happy_flow_-_generate_params
FAIL
``` 

`requiresInterpolation` ends up being `false`, even though there's a templated field in the list generator.

**Comment by crenshaw-dev** [comment]
I'm going to revert this for now, because I think it breaks interpolation for fast-templated nested matrix generators. I think the mocks in the unit tests were hiding the issue from us, but new tests in https://github.com/argoproj/argo-cd/pull/12949/files revealed the issue.

**Comment by wanddynosios** [comment]
Hi there,

just wanted to add that we experienced this (or a very similar issue) with `goTemplate: true`

```yaml
apiVersion: argoproj.io/v1alpha1
kind: ApplicationSet
metadata:
  name: cluster-addons-panda
  namespace: argo-cd
spec:
  generators:
  - matrix:
      generators:
      - clusters:
          selector:
            matchLabels:
              argocd.argoproj.io/secret-type: cluster
              cluster: shared
              installation: panda
              region: eu-central-1
      - merge:
          generators:
          - git:
              files:
              - path: cluster-addons/**/config.yaml
              repoURL: https://my.repo/gitops.git
              revision: main
          - git:
              files:
              - path: variants/{{.metadata.labels.installation}}/**/config.yaml
              repoURL: https://my.repo/gitops.git
              revision: main
          - git:
              files:
              - path: envs/{{.metadata.labels.installation}}/{{.metadata.labels.environment}}/{{.metadata.labels.cluster}}/**/config.yaml
              repoURL: https://my.repo/gitops.git
              revision: main
          mergeKeys:
          - chart_name
    selector:
      matchLabels:
        install: "yes"
  goTemplate: true
  template:
    metadata:
      labels:
        environment: '{{.metadata.labels.environment}}'
        installation: '{{.metadata.labels.installation}}'
        region: '{{.metadata.labels.region}}'
      name: '{{.chart_name}}-{{.name}}'
    spec:
      destination:
        name: '{{.name}}'
        namespace: '{{.chart_name}}'
      project: default
      sources:
      - chart: '{{.chart.project}}'
        helm:
          ignoreMissingValueFiles: true
          valueFiles:
          - $values/cluster-addons/{{.chart_name}}/values.yaml
          - $values/variants/{{.metadata.labels.installation}}/{{.chart_name}}/values.yaml
          - $values/envs/{{.metadata.labels.installation}}/{{.metadata.labels.environment}}/{{.metadata.labels.cluster}}/{{.chart_name}}/values.yaml
        repoURL: '{{.chart.repository}}'
        targetRevision: '{{.chart.version}}'
      - ref: values
        repoURL: https://my.repo/gitops.git
        targetRevision: main
      syncPolicy:
        automated:
          allowEmpty: true
          prune: false
          selfHeal: true
        retry:
          backoff:
            duration: 5s
            factor: 2
            maxDuration: 5m
          limit: 5
        syncOptions:
        - CreateNamespace=true
```

It does not occur with the [latest build](https://quay.io/repository/argoproj/argocd/manifest/sha256:5f3a835028c51c3225054957bbb609c1af449bedd429c71d9b1feb1dc105c7cb).
